### ISSUE TITLE
how to use custom results.yaml?

### ISSUE THREAD

**Issue by martinpitt** [teemtee/tmt#1671]
I am trying to use the new "custom results.yaml" feature from PR #1488 for improving rpminspect testing, in https://github.com/fedora-ci/rpminspect-runner/pull/80 and https://github.com/fedora-ci/rpminspect-pipeline/pull/64 . But I am confused how this should be used, as this whole PR talks about *test* level results.yaml.

As a reference, look at [this rpminspect run](https://artifacts.dev.testing-farm.io/32eccead-7105-4edb-bab3-4660fe895061/), especially its [artifact dir](https://artifacts.dev.testing-farm.io/32eccead-7105-4edb-bab3-4660fe895061/work-rpminspect0a1Ftj/). You see that it has a plan-level [rpminspect/execute/results.yaml](https://artifacts.dev.testing-farm.io/32eccead-7105-4edb-bab3-4660fe895061/work-rpminspect0a1Ftj/rpminspect/execute/results.yaml):
```yaml
/license:
    result: pass
    log:
      - data/license/output.txt
    duration: 00:05:30
# [... more tests ...]
```

But in the [test-level artifacts dir](https://artifacts.dev.testing-farm.io/32eccead-7105-4edb-bab3-4660fe895061/work-rpminspect0a1Ftj/rpminspect/execute/data/license/) there is no results.yaml (not in data/ either, that's empty). This also makes sense as the plan-level results.yaml from above *already* enumerates the tests in the plan and their results.

The above PRs rework rpminspect to create a custom results.yaml in `$TMT_TEST_DATA`, see [this run](https://artifacts.dev.testing-farm.io/acdee065-2e81-4a0a-aa48-1b87c0b85dff/). Its *test level* artifacts in [rpminspect/execute/data/rpminspect/data/](https://artifacts.dev.testing-farm.io/acdee065-2e81-4a0a-aa48-1b87c0b85dff/work-rpminspectePCcoq/rpminspect/execute/data/rpminspect/data/) contain my custom results.yaml:
```yaml
- name: /rpminspect
  result: fail
  log:
    - verbose.log
    - results.json
```

The format of that is of course a little bogus, as this is a single test and thus should not be a list and assign `name:`. But that's the format that the [documentation](https://github.com/teemtee/tmt/pull/1488/files#diff-8a13dad0801becdd846f804a0e0aa44bc6dda38182c1a072113bc0bd31f49486R37) describes.

However, this isn't reflected at all in the plan-level results in [/rpminspect/execute/results.yaml](https://artifacts.dev.testing-farm.io/acdee065-2e81-4a0a-aa48-1b87c0b85dff/work-rpminspectePCcoq/rpminspect/execute/results.yaml):
```yaml
/rpminspect:
    result: fail
    log:
      - data/rpminspect/output.txt
    duration: 00:03:49
```
i.e. the extra log files are not shown at all. Now, the testing farm still runs tmt 1.16 and thus does not contain this PR yet. But I ran it locally with tmt 1.18.0, and the plan-level result in `/var/tmp/tmt/run-001/rpminspect/execute/results.yaml` looks the same, except for some new (and uninteresting/empty) `ids:`
```yaml
/rpminspect:
    result: fail
    log:
      - data/rpminspect/output.txt
    duration: 00:03:01
    ids:
        id:
        extra-nitrate:
        extra-task:
```

The documented custom results.yaml format would actually make much more sense at the plan level, as it enumerates the tests. So I tried that, i.e. [put it into `$TMT_PLAN_DATA`](https://github.com/fedora-ci/rpminspect-runner/pull/80/commits/1c0573501ac96b4d60c3aa8a1d16c01baaf53594). Re-running tmt locally, I get
```
$ find /var/tmp/tmt/run-002 -name results.yaml
/var/tmp/tmt/run-002/rpminspect/data/results.yaml
/var/tmp/tmt/run-002/rpminspect/execute/results.yaml
```

the first one is my custom file, but it's just "data":
```
$ cat /var/tmp/tmt/run-002/rpminspect/data/results.yaml
- name: /rpminspect
  result: fail
  log:
    - data/rpminspect/output.txt
    - verbose.log
    - results.json
```

and the second one is the actual results, but it's just the standard format, without my custom log files:
```
$ cat /var/tmp/tmt/run-002/rpminspect/execute/results.yaml
/rpminspect:
    result: fail
    log:
      - data/rpminspect/output.txt
    duration: 00:03:22
    ids:
        id:
        extra-nitrate:
        extra-task:
```

So, I am none the wiser. It seems to me like #1488 did not change anything user-visible at all both with the documented `${TMT_TEST_DATA}/results.yaml` approach nor with putting it into `$TMT_PLAN_DATA`. How can this be used?

**Comment by lukaszachy** [comment]
@martinpitt  Have you changed your tests' metadata to contain 'result: custom` ?

**Comment by martinpitt** [comment]
Ah, I didn't, thanks! https://tmt.readthedocs.io/en/latest/spec/tests.html#result is a little unclear where I'd set that in our [.fmf file](https://github.com/martinpitt/rpminspect-pipeline/blob/custom-results-yaml/rpminspect.fmf). I messed around with that a little, and this works:

```diff
+++ b/rpminspect.fmf
@@ -6,6 +6,7 @@ discover:
     - name: /rpminspect
       framework: shell
       test: rpminspect_runner.sh $TASK_ID $PREVIOUS_TAG
+      result: custom
       duration: 720m
```

I'm now getting the extra files:
```
$ cat /var/tmp/tmt/run-001/rpminspect/execute/results.yaml
/rpminspect/rpminspect:
    result: fail
    log:
      - data/rpminspect/output.txt
      - verbose.log
      - results.json
    ids:
        id:
        extra-nitrate:
        extra-task:

```

Now I just need to figure out how to turn that `/rpminspect/rpminspect:` into just `/rpminspect:`, but that should be easy :grin: 

Thanks, and sorry for the noise! :bow:

**Comment by martinpitt** [comment]
There's still that structural bit that I don't understand, though: The docs say, that the custom results.yaml should look like this:
```yaml
    - name: /test/passing
      result: pass
      log: pass_log
      duration: 00:11:22
      note: good result
    - name: /test/failing
      result: fail
      log: fail_log
      duration: 00:22:33
      note: fail result
    - name: /test/no_keys
      result: pass
```

But how can that be? This is the result of a single test, how/why does it describe the result of multiple tests? It would seem a lot more logical to me if this was:
```yaml
result: pass
log: pass_log
duration: 00:11:22
note: good result
```

**Comment by martinpitt** [comment]
> I just need to figure out how to turn that /rpminspect/rpminspect: into just /rpminspect:, but that should be easy

FTR, I spent half an hour on this and can't figure it out. I'll just accept it as-is, and hope that it doesn't break anything (it will most probably just look weird)

**Comment by lukaszachy** [comment]
tmt allows multiple results to be reported (which should exactly rpminspect case) so custom results file is list. Having both list or dict accepted would just make it harder to validate.

**Comment by martinpitt** [comment]
OK, thanks! Is there some way to report just a single result, without inventing a new name:? I already tried to drop `name:` (fails with a `KeyError`) or set it to `/` (but then the final name also does, which is weird).

**Comment by martinpitt** [comment]
We actually don't want to split it up, as that would make it a lot harder to find the extra log files and the viewer.html link. We *really* want to report these on the plan level, not test level, but I didn't find a way to do that. So minimizing number of tests and plans to 1 seems to be the best that we can do right now?

**Comment by lukaszachy** [comment]
> We actually don't want to split it up, as that would make it a lot harder to find the extra log files and the viewer.html link. 

Is there something tmt can do better to use the multiple results from single test? I'm not sure why it is harder to find logs as their paths should be copied to final results.yaml

>  Is there some way to report just a single result, without inventing a new name:? 

I'm afraid we have a bug here. I'd say that `name: /` should map to the original name, instead of appending the '/' 

E.g. this should report `pass /foo` but it does report `pass /foo/` 
WDYT @ZelenyMartin ?
```
discover:
  how: shell
  tests:
  - name: foo
    result: custom
    test: |
      cat > ${TMT_TEST_DATA}/results.yaml <<EOF
        - result: pass
          name: /
      EOF
```

**Comment by martinpitt** [comment]
> Is there something tmt can do better to use the multiple results from single test?

No, that's fine. The idea is that we have a [custom rpminspect viewer](https://github.com/rpminspect/rpminspect/pull/905) which we want to expose in the Testing Farm/tmt runs for rpminspect. I.e. the generic TF viewer doesn't need to do any effort. At some point we should directly open the custom viewer, but that will take some steps. Until then, we just want to expose that viewer.html as a test artifact, and make it as easy to find as possible.

**Comment by psss** [comment]
Seems there is nothing left to be still addressed, closing. Feel free to reopen if it is not the case.
